pkg/controller/datavolume/controller.go

```diff
@@ -236,63 +231,81 @@ func getDataVolumeOp(dv *cdiv1.DataVolume) dataVolumeOp {
-	res, syncErr := r.sync(log, req, nil, nil)
-	return r.updateStatus(log, res, syncErr)
-}
-
-func (r ReconcilerBase) sync(log logr.Logger, req reconcile.Request, cleanup, prepare func(*cdiv1.DataVolume) error) (dataVolumeSyncResult, error) {
-	var ctx dataVolumeSyncResult
-
+func (r ReconcilerBase) syncCommon(log logr.Logger, req reconcile.Request, syncRes *dataVolumeSyncResult, cleanup, prepare func() error) error {
 	dv, err := r.getDataVolume(req.NamespacedName)
-	if dv == nil {
-		ctx.res = &reconcile.Result{}
-		return ctx, err
+	if dv == nil || err != nil {
+		syncRes.result = &reconcile.Result{}
+		return err
 	}
+	syncRes.dv = dv
+	syncRes.dvCopy = dv.DeepCopy()
+
 	if dv.DeletionTimestamp != nil {
 		log.Info("DataVolume marked for deletion, cleaning up")
 		if cleanup != nil {
-			err := cleanup(dv)
-			return ctx, err
+			err := cleanup()
+			return err
 		}
-		ctx.res = &reconcile.Result{}
-		return ctx, nil
+		syncRes.result = &reconcile.Result{}
+		return nil
 	}
 
 	if prepare != nil {
-		if err := prepare(dv); err != nil {
-			return ctx, err
+		if err := prepare(); err != nil {
+			return err
 		}
 	}
 
-	pvc, err := r.getPVC(dv)
+	syncRes.pvc, err = r.getPVC(dv)
 	if err != nil {
-		return ctx, err
+		return err
 	}
-	if pvc != nil {
-		res, err := r.garbageCollect(dv, pvc, log)
-		if err != nil {
-			return ctx, err
+	if syncRes.pvc != nil {
+		if err := r.garbageCollect(syncRes, log); err != nil {
+			return err
 		}
-		if res != nil {
-			ctx.res = res
-			return ctx, nil
+		if syncRes.result != nil || syncRes.dv == nil {
+			return nil
 		}
-		if err := r.validatePVC(dv, pvc); err != nil {
-			return ctx, err
+		if err := r.validatePVC(dv, syncRes.pvc); err != nil {
+			return err
 		}
+		r.annotate(syncRes.dvCopy, syncRes.pvc)
 	}
 
-	pvcSpec, err := renderPvcSpec(r.client, r.recorder, log, dv)
+	syncRes.pvcSpec, err = renderPvcSpec(r.client, r.recorder, log, dv)
 	if err != nil {
-		return ctx, err
+		return err
+	}
+	return nil
+}
+
+func (r ReconcilerBase) syncUpdateMeta(log logr.Logger, syncRes dataVolumeSyncResult) error {
+	if syncRes.dv == nil || syncRes.dvCopy == nil {
+		return nil
+	}
+	if !reflect.DeepEqual(syncRes.dv.Status, syncRes.dvCopy.Status) {
+		return fmt.Errorf("status update is not allowed in sync phase")
+	}
+	if !reflect.DeepEqual(syncRes.dv.ObjectMeta, syncRes.dvCopy.ObjectMeta) {
+		if err := r.updateDataVolume(syncRes.dvCopy); err != nil {
+			r.log.Error(err, "Unable to sync update dv meta", "name", syncRes.dvCopy.Name)
+			return err
+		}
+		// Needed for emitEvent() DeepEqual check
+		objMeta := syncRes.dvCopy.ObjectMeta.DeepCopy()
+		syncRes.dv.ObjectMeta = *objMeta
+	}
+	return nil
+}
+
+func (r ReconcilerBase) annotate(dv *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) {
+	if pvc.Status.Phase == corev1.ClaimBound && pvcIsPopulated(pvc, dv) {
+		if dv.Annotations == nil {
+			dv.Annotations = make(map[string]string)
+		}
+		dv.Annotations[cc.AnnPrePopulated] = pvc.Name
```

Comment:
I think it appears that each of the "derived" controllers may add their own annotations so it is okay that they each call `syncUpdateMeta` explicitly but that may not always be the case. And if so, this update would get lost.  Should syncCommon also call `syncUpdateMeta`?

Reply:
It won't get lost as the check in emitEvent() will always error on:
```
	if !reflect.DeepEqual(dataVolume.ObjectMeta, dataVolumeCopy.ObjectMeta) {
		return fmt.Errorf("meta update is not allowed in updateStatus phase")
	}
```

Reply:
This error would not be easy to debug either, as it is returned by `updateStatus` but the metadata is mutated in `sync`.  The peril of having shared state

Reply:
Adding `syncUpdateMeta` call in `syncCommon`.